File: solutions/find-all-numbers-disappeared-in-an-array.cpp
/*
 *Solution of problem 448
 *https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
 */


//efficient approach: O(n) runtime with O(1) space 
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {

        vector<int> ans;
        for(int i=0; i<nums.size(); i++){
            int m = abs(nums[i]) - 1;
            nums[m] = nums[m]>0 ? -nums[m] : nums[m];
        }
        for(int i=0; i<nums.size(); i++){
            if(nums[i] > 0)
                ans.push_back(i+1);
        }
        return ans;
    }
};


/*Trivial approach
 *
 *class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        
        vector<int> ans;
        sort(nums.begin(), nums.end());
        int numsSize = nums.size();
        for(int i=1; i<=numsSize; i++){
            if(not binary_search(nums.begin(), nums.end(), i)){
                ans.push_back(i);
            }
        }
        return ans;
    }
};
*
*/

File: solutions/find-all-numbers-disappeared-in-an-array.py
"""
448. Find All Numbers Disappeared in an Array
- https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
- https://leetcode.com/explore/featured/card/fun-with-arrays/523/conclusion/3270/
"""

# Method 1
# Time: O(N)
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        for i in range(len(nums)):
            index = abs(nums[i]) - 1
            nums[index] = - abs(nums[index])
        return [i + 1 for i in range(len(nums)) if nums[i] > 0]
